app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/offline/OfflineDatabaseManagement.kt

```diff
@@ -0,0 +1,194 @@
+    }
+
+    override suspend fun getAllPictures(category: Category): Set<CategorizedPicture> {
+        require(category is OfflineCategory)
+        return try {
+            databaseService.getAllPictures(category)
+        } catch (e: IllegalArgumentException) {
+            throw e
+        }
+    }
+
+    override suspend fun removeCategory(category: Category) {
+        require(category is OfflineCategory)
+        try {
+            databaseService.removeCategory(category)
+        } catch (e: Exception) {
+            //do nothing since this means that the category is not in the database which is the same as having it removed
+        }
+    }
+
+    override suspend fun removePicture(picture: CategorizedPicture) {
+        require(picture is OfflineCategorizedPicture)
+        try {
+            databaseService.removePicture(picture)
+        } catch (e: Exception) {
+            //do nothing since this means that the picture is not in the database which is the same as having it removed
+        }
+    }
+
+    override suspend fun putDataset(name: String, categories: Set<Category>): Dataset {
+        return databaseService.putDataset(name, categories)
+    }
+
+    override suspend fun getDatasetById(id: Any): Dataset? {
+        require(id is String)
+        return databaseService.getDataset(id)
+    }
+
+    override suspend fun getDatasetByName(datasetName: String): Collection<Dataset> {
+        val datasets = databaseService.getDatasets()
+        val res: MutableSet<OfflineDataset> = mutableSetOf()
+        for (d in datasets) {
+            if (d.name == datasetName) {
+                res.add(d as OfflineDataset)
+            }
+        }
+        return res.toSet()
+    }
+
+    override suspend fun deleteDataset(id: Any) {
+        require(id is String)
+        try {
+            databaseService.deleteDataset(id)
+        } catch (e: IllegalArgumentException) {
+            //do nothing since this means that the dataset is not in the database which is the same as having it removed
+        }
+    }
+
+    override suspend fun putRepresentativePicture(picture: Uri, category: Category) {
+        require(category is OfflineCategory)
+        try {
+            databaseService.putRepresentativePicture(picture, category)
+        } catch (e: IllegalArgumentException) {
+            throw e
+        }
+    }
+
+    override suspend fun getDatasets(): Set<Dataset> {
+        return databaseService.getDatasets()
+    }
+
+    override suspend fun getDatasetNames(): Collection<String> {
+        val datasets = databaseService.getDatasets()
+        val res: MutableSet<String> = mutableSetOf()
+        for (d in datasets) {
+            res.add(d.name)
+        }
+        return res.toSet()
+    }
+
+    override suspend fun getDatasetIds(): Set<Any> {
+        val datasets = databaseService.getDatasets()
+        val res: MutableSet<String> = mutableSetOf()
+        for (d in datasets) {
+            res.add(d.id as String)
```

Comment:
```suggestion
            res.add(d.id)
```
superfluous per compiler warning